Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/Internal/CoreDocumentSynchronizationConfig.swift

```diff
@@ -250,27 +110,80 @@ internal class CoreDocumentSynchronization: Hashable {
          documentId: AnyBSONValue,
-         errorListener: FatalErrorListener?) throws {
+         errorListener: FatalErrorListener?) {
         self.docsColl = docsColl
-        self.config = Config.init(namespace: namespace,
-                                  documentId: HashableBSONValue.init(documentId),
-                                  lastUncommittedChangeEvent: nil,
-                                  lastResolution: 0,
-                                  lastKnownRemoteVersion: nil,
-                                  isStale: false,
-                                  isPaused: false)
+        self.namespace = namespace
+        self.documentId = HashableBSONValue.init(documentId)
+        self.uncommittedChangeEvent = nil
+        self.lastResolution = 0
+        self.lastKnownRemoteVersion = nil
+        self._isStale = false
+        self._isPaused = false
         self.errorListener = errorListener
+        self.docLock = ReadWriteLock(label: "document_lock_\(namespace)_\(documentId.value)")
     }
 
-    init(docsColl: ThreadSafeMongoCollection<CoreDocumentSynchronization.Config>,
-         config: inout Config,
-         errorListener: FatalErrorListener?) throws {
-        self.docsColl = docsColl
-        self.config = config
-        self.errorListener = errorListener
+    init(from decoder: Decoder) throws {
+        let values = try decoder.container(keyedBy: CodingKeys.self)
+
+        // verify schema version
+        let schemaVersion = try values.decode(Int32.self, forKey: .schemaVersion)
+        if schemaVersion != 1 {
+            throw DataSynchronizerError.decodingError(
+                "unexpected schema version \(schemaVersion) for CoreDocumentSynchronization")
+        }
+
+        self.namespace = try values.decode(MongoNamespace.self, forKey: .namespace)
+
+        if let lastKnownRemoteVersion =
+            try values.decodeIfPresent(Document.self, forKey: .lastKnownRemoteVersion) {
+            self.lastKnownRemoteVersion = lastKnownRemoteVersion
+        }
+
+        if let eventBin = try values.decodeIfPresent(Binary.self, forKey: .uncommittedChangeEvent) {
+            let eventDocument = Document.init(fromBSON: eventBin.data)
+
+            self.uncommittedChangeEvent =
+                try BSONDecoder().decode(ChangeEvent.self, from: eventDocument)
+        }
+
+        self.documentId = try values.decode(HashableBSONValue.self, forKey: .documentId)
+        self.lastResolution = try values.decode(Int64.self, forKey: .lastResolution)
+        self._isStale = try values.decode(Bool.self, forKey: .isStale)
+        self._isPaused = try values.decode(Bool.self, forKey: .isPaused)
+        self.docsColl = try values.decode(ThreadSafeMongoCollection.self, forKey: .docsColl)
+        self.docLock = ReadWriteLock(label: "document_lock_\(namespace)_\(documentId.value)")
+    }
+
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
```

Comment:
[major race condition, but might be worth opening a separate ticket] I think you need to take a `docLock.read` lock over this entire method. It would be pretty major race condition if this document config could change while it was being persisted, ending up with a partially corrupt persisted config. This is going to be tricky without reentrant lock semantics so it might be worth opening a ticket to have our locks have working reentrant semantics